keps/sig-node/127-user-namespaces/README.md

```diff
@@ -316,121 +325,121 @@ The picked range will be stored under a file named `userns` in the pod folder
-the same mapping if the pod has other volumes.
+* Changing the owner of files in configmaps/secrets/etc was a concern, those
+  volume type today ignore FsUser setting and changing it to honor them was a
+  concern for some members of the community due to possibly breaking some
+  existing behavior.
 
-While these pods do not use a distinct user namespace mapping, they are still
-using a new user namespace object in the kernel (so they cannot join/attack
-other pods namespaces). Security-wise this is a middle layer between what we
-have today (no userns at all) and using a distinct UID/GID mapping for the user
-namespace.
+* Therefore, we need to rely on `fsGroup`, that always mark files readable for
+  the group. This means this solution can't work when we NEED not to have
+  permissions for the group (think of ssh keys, for example, that tools enforce no
+  permission for the group)
 
-#### Phase 3: TBD
+* There are several other files that also need to have the proper permissions,
+  like: /dev/termination-log, /etc/hosts, /etc/resolv.conf, /etc/hotname, etc.
+  While it is completely possible to do, it adds more complexity to lot of parts
+  of the code base and future Kubernetes features will have to take this into
+  account too. Furthermore, some of these files are created by container runtimes,
+  so complexity creeps out very easily.
 
-#### Unresolved
+##### Example without idmap mounts
 
-Here is a list of considerations raised in PRs discussion that hasn't yet
-settle. This list is not exhaustive, we are just trying to put the things that
-we don't want to forget or want to highlight it. Some things that are obvious we
-need to tackle are not listed. Let us know if you think it is important to add
-something else to this list:
+Now let's say the pod is using its volumes that were mounted using idmap mounts
+by the container runtime. All the mappings used (the idmap mounts and the pod
+userns) are:
+
+| Host ID  | Container ID   | Length  |
+|----------|----------------|---------|
+| 65536    |      0         |  65536  |
 
-- We will start with mappings of 64K. Tim Hockin, however, has expressed
-  concerns. See more info on [this Github discussion](https://github.com/kubernetes/enhancements/pull/3065#discussion_r781676224)
-  SergeyKanzhelev [suggested a nice alternative](https://github.com/kubernetes/enhancements/pull/3065#discussion_r807408134),
-  to limit the number of pods so we guarantee enough spare host UIDs in case we
-  need them for the future. There is no final decision yet on how to handle this.
-  For now we will limit the number of pods, so the wide mapping is not
-  problematic, but [there are downsides to this too](https://github.com/kubernetes/enhancements/pull/3065#discussion_r812806223)
 
+Keep in mind the file `/var/lib/kubelet/.../volumes/configmap/foo` is created by
+the kubelet, the owner is host UID 0, **but** it is bind-mounted into
```

Comment:
What if my container doesn't run as root?  What UID will these configmap files come in as?  I still feel like I am missing something

Reply:
@thockin if your container runs or doesn't run as root, the configmap's files ownership is unchanged. Just as it is without userns. It is completely orthogonal.

So, in this case, the configmap's file will be seen as owned by root inside the container. No matter as which user inside the container is the process running.

What idmap mounts gives us is the IDs translation done by the kernel, where if you have a file owned by UID 0 on the host, and you create an idmap bind-mount of that file into the container using the same mapping as the userns, then you will see it as owned by UID 0 *inside the container*.

Without idmap mounts, you will need to create the file using the host IDs of the userns (explained in other sections why this got messy in the code. Will expand that now and push again)

Reply:
So if I run as UID 1 in my container, the file is still owned by UID 0 - which doesn't actually solve issues like SSH keys.  Or am I still misunderstanding?

Shouldn't this idmapped mount map host(0) to whatever UID this container is running as?
 
I spent some time tinkering but failed to prove it in time.

Reply:
I STRONGLY think that should NOT be done. Let me elaborate.

It is true that if you run as UID 1, you will not be the owner of the files in a secret/configmap/etc. However, that is true without userns too. And I think it is a completely orthogonal problem, completely unrelated to userns.

We should NOT change (except with a VERY good reason) the behavior when using userns. It is NOT expected that if you enable some namespace (userns in this case, but be it any other namespace) that file ownership changes: some files that are normally owned by root now are owned by a completely different heuristic. And it is completely out of scope for userns changes too, it is really not some problem we create when introducing userns nor anything.

If we want that pods that run with UID 1 (or any non 0) can be the owners of files mounted from configmaps, then we want that for any pod (with userns or without). The beauty of using idmap mounts, also, is that if we fix it for pods without userns, then we fix it for pods with userns (as the mapping done by the kernel will do the trick: if the files is created with owner UID 1 in the host, then when userns are in use with idmap mounts, it will be UID 1 inside the container the owner for that mount point exposed to the pod).

But changing file ownership only when userns is in use is kind of coupling two different problems, solving it in a very inconsistent way (only when userns are in use) and a problem that is completely out of scope for introducing userns. A way better solution for that is to have a different KEP, that will solve it for all the cases. And I mentioned, by just solving it for the non userns case, it will automatically work for the userns case too.

So, no, I do not think we want to do _anything_ for that as part of this KEP.

Reply:
> it is really not some problem we create when introducing userns

No, but it seems like an opportunity to fix a long-standing problem.

All of the chown issues remain.  If this can dynamically shift the FS accesses to feel more "natural" - that's a huge win.  I'm not necessarily saying for ALL volumes, but for the "pseudo-volumes" discussed here, it seems like a win.  Is there a downside? 

> If we want that pods that run with UID 1 (or any non 0) can be the owners of files mounted from configmaps, then we want that for any pod (with userns or without)

Yes, but we can't achieve it - it has been reported many times (often in the context of SSH) and we just don't have a good answer.

> The beauty of using idmap mounts, also, is that if we fix it for pods without userns, then we fix it for pods with userns

It's not for lack of want that we have not fixed it - we don't have a mechanism because a volume can be shared by N containers in a pod, and those can all be different UIDs.

Reply:
@thockin Can I add an unresolved issue for this, and we discuss it for the next releases (just not 1.27)?

IMHO, it is still an alpha feature, we can do these improvements later, don't need to halt this.

> It's not for lack of want that we have not fixed it - we don't have a mechanism because a volume can be shared by N containers in a pod, and those can all be different UIDs.

Then, how is it that you propose to solve it for pods with userns exactly?

Reply:
I'm not sure there is something we can do, as the file is already created with UID 0 on the host and shared across all containers.

In any case, it seems too late to include this (besides not having a concrete idea, I'm not sure this can work), this would need discussion with sig-storage that meets bui-weekly on Thusday and they don't meet this week.

Can we defer this to another release and still move with the rest of the KEP? Does this LGTY?

(sorry for the late response, it is 10:30pm here).

Reply:
If we know we run as 1234 in container, and we know we write projected files as 0 on the host, I was hoping there was a way where, instead of mapping host 0 -> container 0, we map host 0 -> container 1234.

Please proceed.  My question here is not blocking.  I think this is one of the incentives to run as root-in-container, so this KEP won't fix that, but that's less dire with userns in play.

Reply:
Thanks!

Well, it is not that we map host  --> container 0 directly. We have the userns map and the bind-mount map, and the end result is the combination of both mappings. So it is a little bit more involved. I'll think what we can do for future iterations of this KEP regarding this

I'd like to understand in detail, though, what you meant with your message here: https://github.com/kubernetes/enhancements/pull/3811#discussion_r1097643565

Can you elaborate on the problem you want to see if we can solve/lessen? And why can't we offer good answers and what are the things that we explored and didn't work?

I'd like more context to think about proper solutions and probably reference your answer in the KEP too ;)

Reply:
> Well, it is not that we map host --> container 0 directly.

IIUC we map (for example) container [0-1000] to host [100000-101000] with userns and then map host [0-1000] to host [100000-101000] with idmap.  Right?

We know that, for pseudo-volumes we write files as host 0.

If the container runs as UID 1234, why not map host 0 to host 101234 (100K for that pod's range base + 1234 for that container's UID).  We only need that one mapping for these volumes.

It's equivalent to creating a copy of the pseudo-volume for each container, and chown'ing for just that one.

Again, I tried to prove it works (or doesn't) but I couldn't figure how to make the idmapped mount happen and had to move on to other KEPs :)

Reply:
> Can you elaborate on the problem you want to see if we can solve/lessen? 

We know there are things out there that don't like their files to be owned by a different UID and have group-level access.  It's well-intentioned, usually, but hard to enact when a pod might have N containers running as different UIDs.

fsGroup was always kind of a hack - it would be better if those files APPEARED to be owned by the container UID - at least for pseudo-volumes, maybe for all.

Reply:
@thockin but what can't we improve for pods without userns?

For example, as we do have the mode per key and `defaultMode` for these volumes, why not add a UID/GID to the pod.spec? What is the issue about doing that?

Sure, that won't allow one volume to be mounted as different UIDs in different containers, but if you need it on different containers with different UIDs, it is technically possible to create different volumes (from the same secret) and use different UIDs. That is a _big_ advantage compared to what we have today.

What is the problem with that kind of solutions (besides being kind of clumsy)? We can improve from there to remove the clumsiness. (like add a special value to UID, like "use container RunAsUser").

btw, to play with idmap mounts, you can use the mount command from git or https://github.com/brauner/mount-idmapped. But be aware to play with and without userns (you can create one easily with the `unshare` command).

I'll try to think about this in more detail, but after code freeze and all of that.

Reply:
> but why can't we improve for pods without userns?
>
> For example, as we do have the mode per key and defaultMode for these volumes, why not add a UID/GID to the pod.spec? What is the issue about doing that?

The "different containers" problem is a real one and shouldn't be dismissed.  You're right that simple chown might be enough for most people - we haven't pursued it because it's imperfect and hasn't risen to the top of anyone's list (workarounds exist for most cases).  My request here was more "opportunistic" than urgent.

>  to play with idmap mounts, you can use the mount command from git or https://github.com/brauner/mount-idmapped. 

I could not get it to work and it wasn't clear why, so I moved on.  I suspected we would not undetake this idea at the same time as userns, so it was more "why can't we" than "we need to". :)